### PR TITLE
Webview async resolution, take 2

### PR DESCRIPTION
**Pull request by christian-bromann**
branched off https://github.com/stateful/tangle/pull/41

### CODE REVIEW

**Comment by christian-bromann** [comment]
Verified functionality within our awesome-ux project, see https://github.com/stateful/vscode-awesome-ux/pull/18. Will check out our internal Stateful app next.

**Comment by christian-bromann** [comment]
@sourishkrout please review latest commits. I made it so that we don't re-emit for states that don't change. Tested it in `vscode-awesome-ux` and worked as expected. This should help improve performance in Marquee as well. The implementation can probably be improved to be more rxjs like, maybe you have an idea.

**Comment by sourishkrout** [comment]
> @sourishkrout please review latest commits. I made it so that we don't re-emit for states that don't change. Tested it in `vscode-awesome-ux` and worked as expected. This should help improve performance in Marquee as well. The implementation can probably be improved to be more rxjs like, maybe you have an idea.

For the re-mission of states if they don't change: Just this commit https://github.com/stateful/tangle/pull/43/commits/44960d4cc4d367b231ac3e6bfca283e8d7799b49 right? Just trying to isolate the diff to reload the context into my brain.

**Comment by christian-bromann** [comment]
Tested on [`stateful/app`](https://github.com/stateful/app/pull/605) and everything seems fine. Next project: Marquee.

**Comment by christian-bromann** [comment]
> Just this commit [44960d4](https://github.com/stateful/tangle/commit/44960d4cc4d367b231ac3e6bfca283e8d7799b49) right?

Correct!

**Comment by christian-bromann** [comment]
[Tested it](https://github.com/stateful/vscode-marquee/pull/234) successfully in Marquee. Preventing listener calls on states that haven't changed helped improve the performance.

File: src/tangle.ts [hunk 1]
```diff
@@ -124,11 +136,20 @@ export class Client<T> {
...
+            filter(value => typeof value !== 'undefined'),
+            map(value => value as T[K])
+        ).subscribe((val: T[K]) => {
+            if (val === lastValue) {
+                return;
+            }
+
+            lastValue = val;
+            fn(val);
+        });
```
Comment (sourishkrout) [inline src/tangle.ts:152]: You can simplify this with a rxjs-native operator like so:

```suggestion
            map(value => value as T[K])
            distinctUntilChanged()
        ).subscribe(fn);
```

**Review by sourishkrout — APPROVED** [review]
👍 sounds like you did enough testing... let's merge after making the slight change commented on above!

File: src/tangle.ts (hunk 1 above)
Comment (christian-bromann) [reply src/tangle.ts:152]: This doesn't seem to work. It seems that the initial/first change is not propagated, e.g. this test fails:

```ts
const namespace = 'test1';

    const ch = new Channel<Payload>(namespace, { someProp: 0 } as any);
    const bus = await ch.registerPromise([
        new Worker(workerPath, { argv, workerData: { channel: namespace, add: 1 } })
    ]);
    let result = 0;
    bus.listen('someProp', (sum?: number) => {
        result += sum || 0;
    });

    bus.broadcast({ someProp: 1 });
    bus.broadcast({ someProp: 1 });
    bus.broadcast({ someProp: 1 });
    bus.broadcast({ someProp: 1 });
    bus.broadcast({ someProp: 1 });
    bus.broadcast({ someProp: 1 });
    // implicitly wait to allow events to come in
    await new Promise((resolve) => setTimeout(resolve, 200));

    expect(result).toBe(1);
    ch.providers.map((p) => p.terminate());
```

The state change listener is not called once even though it should at the first time where state changes from `0` to `1`. This causes this and other test to fail and is not the behavior I would expect.

Comment (sourishkrout) [reply src/tangle.ts:152]: Can you commit/push the changes you made to the branch please? The ones that make the test case fail. I can't reproduce the behavior with `distinctUntilChanged()`. Figured it'd be easier to troubleshoot off the same branch.

Comment (sourishkrout) [reply src/tangle.ts:152]: @christian-bromann maybe i'm overlooking something

**Comment by christian-bromann** [comment]
@sourishkrout pushed the change, see last build is failing now.

File: src/tangle.ts [hunk 2]
```diff
@@ -136,20 +137,13 @@ export class Client<T> {
...
-        ).subscribe((val: T[K]) => {
-            if (val === lastValue) {
-                return;
-            }
-
-            lastValue = val;
-            fn(val);
-        });
+            map(value => value as T[K]),
+            distinctUntilChanged((prev, curr) => prev === curr)
```
Comment (sourishkrout) [inline src/tangle.ts:145]: you can just do `distinctUntilChanged()` since the default comparator will do the job.

File: src/tangle.ts [hunk 3]
```diff
@@ -136,20 +137,13 @@ export class Client<T> {
...
+        ).subscribe();
```
Comment (sourishkrout) [inline src/tangle.ts:146]: you'll need to pass a subscriber into the subscription which is why no values are being emitted:

`).subscribe(fn);` should make it work.

**Review by sourishkrout — CHANGES_REQUESTED** [review]

File: src/tangle.ts (hunk 2 above)
Comment (sourishkrout) [reply src/tangle.ts:145]: Btw, if we want complex types (ref vs val) other than numbers and strings to work we’d need a JSON stringify comparator which could get expensive unless we really need it.

File: src/tangle.ts (hunk 3 above)
Comment (christian-bromann) [reply src/tangle.ts]: dang, this did the trick

**Comment by christian-bromann** [comment]
Good to go now?

**Review by sourishkrout — APPROVED** [review]

**Comment by sourishkrout** [comment]
LGTM 👌🏼

**Pull request merged**